Remove PathTooLongException and associated tests.

stephenmichaelf opened the pull request:
Remove preemptive check for path too long. Refer to #8655 .

@JeremyKuhne @karelz

karelz commented on src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs:
```diff
@@ -206,19 +206,6 @@ public void DirectoryEqualToMaxDirectory_CanBeCreatedAllAtOnce()
-                Assert.Throws<PathTooLongException>(() => Create(path));
-            });
-        }
-
```
I think we should keep a test which checks the long path (even though it is "variable" and not known ahead of time).

karelz replied:
The exception should be compatible with Desktop for compat.

JeremyKuhne replied:
> The exception should be compatible with Desktop for compat.

It can't be by design.

Windows doesn't always return "TooLong". The whole point of this is to not second-guess the OS. We already get this wrong for some file system formats and we're currently blocking any new ones that might support > 255.

JeremyKuhne replied:
"Compat" with this change is we should get an IOException of some sort.

stephenmichaelf replied:
To clarify for this test, are you saying we should add the test back and just expect a different type of IOException?

JeremyKuhne replied:
Yes- these exceptions may change, but as long as they're some type of IOException we should be ok.

karelz replied:
That means we should have tests growing the path length until an IOException happens, right?

JeremyKuhne commented on src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs:
```diff
@@ -209,17 +209,6 @@ public void SearchPatternIgnoreSubDirectories()
-        [PlatformSpecific(TestPlatforms.Windows)]
-        public void WindowsSearchPatternLongSegment()
-        {
-            // Create a path segment longer than the normal max of 255
```
Again, on this one I don't care what it throws, just as long as it fails with some sort of IO exception.

JeremyKuhne commented on src/System.IO.FileSystem/tests/Directory/Move.cs:
```diff
@@ -127,18 +127,6 @@ public void IncludeSubdirectories()
-        {
-            string testDir = GetTestFilePath();
-            Directory.CreateDirectory(testDir);
-            Assert.All((IOInputs.GetPathsLongerThanMaxLongPath(GetTestFilePath())), (path) =>
```
This should continue to throw in some way. I would expect either "too long" or "not found" or possibly "invalid" for any of these long paths. It depends on where they get kicked back in the system.

JeremyKuhne commented on src/System.IO.FileSystem/tests/File/Create.cs:
```diff
@@ -149,19 +149,6 @@ public void OverwriteReadOnly()
-            DirectoryInfo testDir = Directory.CreateDirectory(GetTestFilePath());
-            Assert.Throws<PathTooLongException>(() => Create(Path.Combine(testDir.FullName, new string('a', 300))));
-
-            //TODO #645: File creation does not yet have long path support on Unix or Windows
```
This is a weird commented out blob. It should just be deleted- it should fail with segments that are this long.

JeremyKuhne commented on src/System.IO.FileSystem/tests/File/Create.cs:
```diff
@@ -149,19 +149,6 @@ public void OverwriteReadOnly()
-        public void LongPath()
-        {
-            DirectoryInfo testDir = Directory.CreateDirectory(GetTestFilePath());
-            Assert.Throws<PathTooLongException>(() => Create(Path.Combine(testDir.FullName, new string('a', 300))));
```
Again, should throw something.

JeremyKuhne commented on src/System.IO.FileSystem/tests/File/Move.cs:
```diff
@@ -176,22 +176,6 @@ public void MaxPath_Windows()
-
-            Assert.All(IOInputs.GetPathsLongerThanMaxLongPath(GetTestFilePath()), (path) =>
-            {
-                Assert.Throws<PathTooLongException>(() => Move(testFileSource, path));
```
Again, some exception

stephenmichaelf replied:
For PathHelpers.Windows.cs I only saw it in one place and that change is in here:

![image](https://cloud.githubusercontent.com/assets/450490/20776824/f03b6512-b730-11e6-923f-2013d5358bd4.png)

I made the other changes and I am testing them now.

JeremyKuhne replied:
https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/IO/PathHelper.Windows.cs#L72 etc.

JeremyKuhne commented on src/System.IO.FileSystem/tests/Directory/CreateDirectory.cs:
```diff
@@ -212,18 +212,18 @@ public void DirectoryWithComponentLongerThanMaxComponentAsPath_ThrowsPathTooLong
         }
 
-        #endregion
+#endregion
```
You've lost some indenting here apparently

JeremyKuhne commented:
What errors did you actually get back? Specifically in [`GetExceptionForWin32Error`](https://github.com/dotnet/corefx/blob/c55435befeb372707558a5c441b3c64bf243948f/src/Common/src/System/IO/Win32Marshal.cs#L47)?

We should validate the HResult is what we expect and/or if it is something "new" that fits one of the existing cases we might want to add it. Specifically I'm wondering if `ERROR_BUFFER_OVERFLOW` is what we're seeing (206)- which we should probably add as another case for PathTooLong.

JeremyKuhne commented:
Looking through things a bit, one other one I would expect is `ERROR_BAD_PATHNAME` (161). That one should probably stay as an IOException.

stephenmichaelf commented:
I am trying to debug in VS but having issues. I keep getting inconsistent results, I can do the following:

```
msbuild System.IO.FileSystem.Tests.builds
```

I used to be able to run the main build:

```
build-managed
```

But now it's failing randomly with:

```
C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Platforms\x64\PlatformToolsets\v140\Toolset.targets(36,5): error MSB8036: The Windows SDK version 8.1 was not found. Install the required version of Windows SDK or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solution". D:\OpenSource\corefx\src\System.ServiceProcess.ServiceController\tests\System.ServiceProcess.ServiceController.TestNativeService\System.ServiceProcess.ServiceController.TestNativeService.vcxproj]
Command execution failed with exit code 1.
```

Then in Visual Studio I can build the solution fine:

```
========== Build: 2 succeeded, 0 failed, 2 up-to-date, 0 skipped ==========
```

But I am seeing "errors" in the files and I see nothing in Test Explorer:

![image](https://cloud.githubusercontent.com/assets/450490/20796415/fec9ec34-b7a3-11e6-9ba6-ef1ed7021c0b.png)

![image](https://cloud.githubusercontent.com/assets/450490/20796436/0e74541c-b7a4-11e6-9db3-cd49f57c0b09.png)

I am using Debug build:

![image](https://cloud.githubusercontent.com/assets/450490/20796444/15a58cec-b7a4-11e6-9147-e6333a00b048.png)

I looked and it seems like the xUnit test runner VSIX is deprecated and that it should be finding my tests. Any ideas what I am doing wrong?

JeremyKuhne commented:
Sorry about the delayed response, everyone is OOF with an office move.

Unfortunately our debugging experience / instructions are less than optimal at the moment. @weshaggard, am I missing any documentation on this that you are aware of? I'm going to open some issues on documenting and improving the experience (particularly from the IDE).

Use the `Windows_Debug` config, you'll see the intellisense/syntax highlighting light up correctly.

The Test Explorer is not currently usable with CoreFX projects. Right click on the test project you want to debug and set it as the startup project. The filtering isn't quite right so you'll need to tweak things a bit before you can run.

If you build the relevant test .builds from the command line you'll get a dump something like the following:

```
  Executing in P:\repos\corefx\bin\tests\Windows_NT.AnyCPU.Debug\System.Runtime.Extensions.Tests\default.netcoreapp1.1\
  Hard linking dependent files...
  Finished linking needed files, moving to running tests.
  Running tests... Start time: 10:58:09.16
  Command(s):
  call CoreRun.exe xunit.console.netcore.exe System.Runtime.Extensions.Tests.dll  -xml testResults.xml -notrait Benchmark=
  true -notrait category=nonnetcoreapp1.1tests  -notrait category=OuterLoop -notrait category=failing -notrait category=no
  nwindowstests
  xUnit.net console test runner (64-bit .NET Core)
  Copyright (C) 2014 Outercurve Foundation.
```

In the debug tab for the properties for the test project you'll see these values. You'll want to tweak the existing command line arguments from

```
xunit.console.netcore.exe System.Runtime.Extensions.Tests.dll -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapp1.1tests -wait -parallel none
```

to whatever you get from the command prompt plus `-wait -parallel none`:

```
xunit.console.netcore.exe System.Runtime.Extensions.Tests.dll  -xml testResults.xml -notrait Benchmark=true -notrait category=nonnetcoreapp1.1tests  -notrait category=OuterLoop -notrait category=failing -notrait category=nonwindowstests -wait -parallel none
```

Then you'll be able to set breakpoints and F5 to run/debug. Note that I'm not sure why it won't run the Windows specific tests, I'll look more next week. In the meantime, just pull the attribute while you're testing.

The [developer guide](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md) has instructions that you can go through.

danmosemsft commented:
This is how I debug tests. Make sure VS is on the path (eg use a developer command prompt) then:

msbuild thetestproject.csproj /t:buildandtest **/p:testdebugger=devenv.exe**

When Xunit starts it will pop the debugger. First time around it will not guess the correct debug engine, you must open solution properties and change it to the Core debug engine (I am using VS2015)  Now set your breakpoints and hit F5.

You can Save the solution and open that solution next time and F5 it will just work without need to reset debugger engine or breakpoints.

stephenmichaelf commented:
@JeremyKuhne  thanks for taking the time to write that up, I appreciate it.

Setting the build to Windows_Debug did fix the intellisense issue.

I also changed the project build tab setting and that is letting me debug in VS.

One strange thing I am seeing is that when exceptions are thrown, that I believe are supposed to be thrown, the debugger is displaying and highlighting them. If I continue it's fine and the tests continue to run. Not sure why that is happening. I checked my exception settings and this is what I have:

![image](https://cloud.githubusercontent.com/assets/450490/20867536/4c19ad9e-ba15-11e6-995b-d2e2bc3586be.png)

I also see this option:

![image](https://cloud.githubusercontent.com/assets/450490/20867551/7fed62b4-ba15-11e6-8b04-b382aa6fd2f7.png)

But it seems problematic for times when I want to see that this type of exception was thrown when it shouldn't be. This is making it super slow to debug :) Is there a setting I need to change?

Here are some of the Win32Errors I am seeing(sorry I missed the path on the first two):

= Directory/GetFileSystemEntries_str_str.cs
WindowsSearchPatternLongSegment()
errorCode = 183 = 0xB7 = ERROR_ALREADY_EXISTS -> So it's throwing an IOException since it goes to default case.
errorCode = 87 = 0x57= ERROR_INVALID_PARAMETER -> Throwing IOException under case statement.

Thanks again for taking the time to walk me through this.

stephenmichaelf commented:
One more question.

What is the process for debugging on other OSes? I see there is a failure in Jenkins:

```
/mnt/j/workspace/dotnet_corefx/master/ubuntu14.04_release_prtest/Tools/tests.targets(257,5): error : One or more tests failed while running tests from 'System.IO.FileSystem.Tests' please check /mnt/j/workspace/dotnet_corefx/master/ubuntu14.04_release_prtest/bin/tests/Unix.AnyCPU.Release/System.IO.FileSystem.Tests/default.netcoreapp1.1/testResults.xml for details! [/mnt/j/workspace/dotnet_corefx/master/ubuntu14.04_release_prtest/src/System.IO.FileSystem/tests/System.IO.FileSystem.Tests.csproj]
```

But I would guess we can't just remote in and check that XML file :)

danmosemsft commented:
@alexperovich can you point to your instructions for debugging a unit test on unix?

karelz commented:
@alexperovich ping? @JeremyKuhne can you please help? The PR is stuck for 2 weeks :(

karelz commented:
@alexperovich @JeremyKuhne ping - the PR is stuck for 1 month :(

alexperovich commented:
@stephenmichaelf @danmosemsft Sorry for the delay. I was on vacation. Instructions for debugging on unix are here: https://github.com/dotnet/corefx/blob/master/Documentation/debugging/unix-instructions.md

JeremyKuhne commented:
@dotnet-bot test this please

JeremyKuhne commented:
Kicking the CI again as the results were deleted.

@stephenmichaelf Sorry about the holiday delays. For first chance exceptions you want to have them off generally- this case in particular is a pain as the ones you care about are hit all the time. What I would do is set a breakpoint in the code that converts the windows error and at the beginning of the tests you care about. Disable/Enable the the error conversion breakpoint when you enter/leave the test you care about.

You can also specify arbitrary tests to run in the command line. [This link](https://github.com/dotnet/buildtools/blob/master/Documentation/test-targets-usage.md) has more details.

stephenmichaelf commented:
@karelz @alexperovich @JeremyKuhne Thanks everyone for the help! I will take a look at this again.

stephenmichaelf commented:
Thank you for the advice it was very helpful, I am now able to run the tests one by one from VS.

@JeremyKuhne  I was debugging the other OSes and I see that the failures in Unix are being throw by:

```
FileSystem.Current.MoveFile()
```

Which throws and then checks /Common/Interop/Unix/Interop.IOErrors.cs:

```
internal static Exception GetExceptionForIoErrno(ErrorInfo errorInfo, string path = null, bool isDirectory = false)
{
	...
	case Error.ENAMETOOLONG:
		return new PathTooLongException(SR.IO_PathTooLong);
	...
}
```

Should I get rid of that case and let it fall through to the default which throws an IOException? Is that the behavior we are looking for?

I was doing more debugging and I found more places that the PathTooLongException is being thrown in PathHelper.Windows.cs and removed them.

After I removed them I started to get unhandled exceptions there:

An exception of type 'System.IO.PathTooLongException' occurred in system.runtime.extensions.dll but was not handled in user code.

It doesn't seem like these are being caught and converted now but I was under the impression we should remove all PathTooLong throws that aren't in conversion code.

I will now check the codes they are throwing in GetExceptionForWin32Error.

Am I making this change harder than it has to be or is it just tricky? :)

karelz commented:
@JeremyKuhne was OOF sick today. CC @ianhays if he can help ...

JeremyKuhne commented:
> Should I get rid of that case and let it fall through to the default which throws an IOException? Is that the behavior we are looking for?

That is the Unix specific helper for errors and shouldn't be changed. We want to surface PathTooLong if the OS tells us that the path is too long. `Win32Marshal.GetExceptionForWin32Error()` is the Windows equivalent.

The Path class code moved to [CoreCLR](https://github.com/dotnet/coreclr/tree/master/src/mscorlib/corefx/System/IO) you'll need to update Path there. The only reason I didn't delete it here is that CoreRT (the WinRT CoreCLR) doesn't have the equivalent move (yet). I've put up a PR to rename these files to reflect this better (they have .uap now).

danmosemsft commented:
@stephenmichaelf do you plan to continue in the manner @JeremyKuhne proposes?

stephenmichaelf commented:
@danmosemsft @JeremyKuhne Yes I can do that.

Just to clarify I should:
1. Get all of the codes for GetExceptionForWin32Error and list them here
2. Remove PathTooLongException from the Path object in CoreCLr
3. Anything else?
4. How do we want to handle the breaking tests in non-Windows? They are still throwing PathTooLongException but we updated the Windows tests(which are passing) to expect an IOException now.

Thanks!

JeremyKuhne commented:
> Get all of the codes for GetExceptionForWin32Error and list them here

I'd add `ERROR_BUFFER_OVERFLOW` (206) in that code as another case for PathTooLong. To get the PathTooLong exceptions you might need to create a `> short.MaxValue` character path to get that to happen for `GetFullPath`.  You should be able to contrive a test case that causes PathTooLong for the other APIs- they might need to have an actual path created that is `< short.MaxValue` with a final file/directory name that pushes the path over.

> Anything else?

If you add `ERROR_BUFFER_OVERFLOW` you'll need to do the same in the helper in CoreCLR- `__Error` is same code. (Btw, feel free to make the name of the helper match CoreFX- we should ultimately share the code here.)

karelz commented:
@stephenmichaelf was the previous answer sufficient?

stephenmichaelf commented:
@karelz Yes I think it is, thanks! I will have a push for this tomorrow.

karelz commented:
@stephenmichaelf ping?

karelz commented:
@stephenmichaelf ping again?

danmosemsft commented:
@stephenmichaelf would love to see this go in, we get a lot of feedback about long path limits...

danmosemsft commented:
I'm going to close this one as it's been a month. We can certainly reopen if anyone would like to finish it off.

karelz commented:
BTW: I unassigned the bug. If anyone wants to finish it off later (incl. @stephenmichaelf), please let us know on the main issue #8655.
